internal/graph/check.go

```diff
@@ -614,6 +614,238 @@ func (c *LocalChecker) ResolveCheck(
+			continue
+		}
+
+		if !condEvalResult.ConditionMet {
+			continue
+		}
+
+		usersetObject, usersetRelation := tuple.SplitObjectRelation(t.GetUser())
+		reqTupleKey := req.GetTupleKey()
+
+		// if the user value is a typed wildcard and the type of the wildcard
+		// matches the target user objectType, then we're done searching
+		if tuple.IsTypedWildcard(usersetObject) && typesystem.IsSchemaVersionSupported(typesys.GetSchemaVersion()) {
+			wildcardType := tuple.GetType(usersetObject)
+
+			if tuple.GetType(reqTupleKey.GetUser()) == wildcardType {
+				span.SetAttributes(attribute.Bool("allowed", true))
+				response.Allowed = true
+				return response, nil
+			}
+
+			continue
+		}
+
+		if usersetRelation != "" {
+			tupleKey := tuple.NewTupleKey(usersetObject, usersetRelation, reqTupleKey.GetUser())
+			handlers = append(handlers, c.dispatch(ctx, req, tupleKey))
+		}
+	}
+
+	if len(handlers) == 0 && errs != nil {
+		telemetry.TraceError(span, errs)
+		return nil, errs
+	}
+
+	resp, err := union(ctx, c.concurrencyLimit, handlers...)
+	if err != nil {
+		telemetry.TraceError(span, err)
+		return nil, errors.Join(errs, err)
+	}
+
+	resp.GetResolutionMetadata().DatastoreQueryCount += response.GetResolutionMetadata().DatastoreQueryCount
+
+	return resp, nil
+}
+
+// checkUsersetPublicWildcardFastPath is the fast path to evaluate userset.
+// The general idea of the algorithm is that it tries to find intersection on the objects as identified in the userset
+// with the objects the user has the specified relation with.
+// For example, for the following model, for check(user:bob, viewer, doc:1)
+//
+//	type group
+//	  define member: [user]
+//	type doc
+//	  define viewer: [group#member]
+//
+// We will first look up the group(s) that are assigned to doc:1
+// Next, we will look up all the group where user:bob is a member of.
+// Finally, find the intersection between the two.
+// To use the fastpath, we will need to ensure that the userset and all the children associated with the userset are
+// exclusively directly assignable. In our case, group member must be directly exclusively assignable.
+func (c *LocalChecker) checkUsersetPublicWildcardFastPath(ctx context.Context, iter storage.TupleKeyIterator, req *ResolveCheckRequest) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "checkUsersetPublicWildcardFastPath")
+	defer span.End()
+	var errs error
+
+	typesys, ok := typesystem.TypesystemFromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("typesystem missing in context")
+	}
+
+	ds, ok := storage.RelationshipTupleReaderFromContext(ctx)
+	if !ok {
+		return nil, fmt.Errorf("relationship tuple reader datastore missing in context")
+	}
+
+	// usersetsMap is a map of all ObjectRelations and its IdsFor example,
+	// [group:1#member, group:2#member, group:1#owner, group:3#owner] will be stored as
+	// [group#member][1]
+	// [group#owner][1, 3]
+	usersetsMap := make(map[string]map[string]struct{})
+
+	for {
+		t, err := iter.Next(ctx)
```

Comment:
If the underlying iterator has to do more than 1 database trip, you will have to wait for all those trips to complete before you begin processing line 797. Suggestion:  put both pieces in different goroutines. Goroutine A sends objectIDs through a map of channels (one per object#relation)

I say leave this as a follow-up because by doing this we open a small can of worms (race conditions, possibility of leaking goroutines etc)

Reply:
Yep, aware of that. In fact, we could start executing the subsequent ones once we observe a whole new ObjectRelation (since the results come sorted) so we could in fact process as a stream. For the time being (this first pass) I'd like to land as is and we can optimize later since I don't envision that to be trivial 😅

Reply:
Adding a NOTE regarding this conversation so its not overlooked in the future.

Reply:
https://github.com/openfga/openfga/pull/1746